plugins/destination/postgresql/client/insert.go

```diff
@@ -59,27 +87,61 @@ func (c *Client) InsertBatch(ctx context.Context, messages message.WriteInserts)
 			return fmt.Errorf("table %s not found", tableName)
 		}
 
-		sql, ok := queries[tableName]
+		detail, ok := details[tableName]
 		if !ok {
+			var sql string
 			// cache the query
 			table := c.normalizeTable(msg.GetTable())
 			if len(table.PrimaryKeysIndexes()) > 0 {
 				sql = c.upsert(table)
 			} else {
 				sql = c.insert(table)
 			}
-			queries[tableName] = sql
+			detail = tableDetail{
+				indexes: table.PrimaryKeysIndexes(),
+				sql:     sql,
+				values:  make(map[string]bool),
+			}
+			details[tableName] = detail
 		}
 
 		rows := c.transformValues(r)
-		for _, rowVals := range rows {
-			batch.Queue(sql, rowVals...)
+		var pkVal []string
+		if _, ok := details[tableName]; ok {
+			pkVal = getPKValues(r, detail.indexes)
+		}
+		for i, rowVals := range rows {
+			if len(detail.indexes) == 0 {
+				batch.Queue(detail.sql, rowVals...)
+				continue
+			}
+			// If the PK value is not in the map of existing pk values, add it to the map and batch the query
+			if _, ok := detail.values[pkVal[i]]; !ok {
+				detail.values[pkVal[i]] = true
+				batch.Queue(detail.sql, rowVals...)
+			} else {
+				c.logger.Warn().Msgf("Duplicate primary key value found for table %s in batch. Flushing batch and creating a new one", tableName)
```

Comment:
I think we don't need a warning message, as this isn't a problem users of the postgres plugin need to concern themselves with and can happen during normal operations; a debug-level message would be better in my opinion.